README.md
# Ghada Express Profiler

Ghada Express Profiler is inspired by [Symfony Profiler][URL_symfony].
Its main target is to make debugging easier for  applications built using [Express Node.js framework][URL_express].

## Works with

- `Express` version `4`

## Dependencies

```json
"ejs": "^3.1.6",
"express": "^4.17.1",
"on-finished": "^2.3.0",
"response-time": "^2.3.2",
"sqlite3": "^5.0.2",
"uuid": "^8.3.2"
```

## Installation

Using npm

```sh
npm install @ghadautopia/express-profiler
```

Or using yarn

```sh
yarn add @ghadautopia/express-profiler
```

## Usage

Import `@ghadautopia/express-profiler` in your project. Instantiate it directly after mounting all the routes in the Express app and before the __404 and Error handlers__. As shown in the below example:

```javascript
// import express
const express = require('express');
// import profiler
const { profiler } = require('@ghadautopia/express-profiler');

// instantiating express app
const app = express();

// app routes
app.use('/', indexRouter);

// instantiating profiler - directly after mounting app routes
profiler(app);

// catch 404 and forward to error handler
app.use(function(req, res, next) {
  next(createError(404));
});

// error handler
app.use(function(err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.render('error');
});
```

## How it works

Ghada Express Profiler mainly depends on 2 concepts:

- __Streams__: responsible for logging the helpful data related to each request
- __Scopes__: responible for using the logged data by streams to render friendly views to be helpful in debugging the request and its response

Ghada Express Profiler uses sqlite to log the `Streams` data. The sqlite file is created in the project itself in the path `.cache/ghada/profiler.sqlite`.

In order to avoid logging collision between concurrent requests. Ghada Express Profiler forces the requests to be handled one by one. If a request is being processed, any other coming request will wait for the current one to finish.

To activate Ghada Express Profiler debugging on a current request, you should prefix the path being called with `/_debug`. For example,

- `http://localhost:3000` should be `http://localhost:3000/_debug`
- `http://localhost:3000/error` should be `http://localhost:3000/_debug/error`

If your app returns HTML content in the response to this request, Ghada Express Profiler will render a __Toolbar__ element sticking to the end of the rendered page. As shown below

![toolbar view](doc/images/view-toolbar.png)

For Ghada Express Profiler to distinguish between requests, it injects 2 headers in the response:

- The `X-Profiler-Link` header: holds the URL of the current request profiler scopes view
- The `X-Profiler-Token` header: holds the profiler token identifier of the current request

![response headers](doc/images/response-headers.png)

By heading to the URL of the `X-Profiler-Link`, You will be able to see something like the below view. This is called the `s` view. On the left hand side navigation, you can check your scopes. Here you can see only the `req-res` scope which is auto imported by the profiler. In the `req-res`, you can check the data about the server request and its response.

![scope request/response](doc/images/view-scopes.png)

To check all the logged requests. Click on the logo anchor on the top left of the page. You should see something like the below view

![scope main](doc/images/view-scopes-main.png)

## Configuration

You can configure Ghada Express Profiler by passing a second argument with the configuration object, as follows:

```javascript
profiler(app, {
  debugRoute: '_myDebugRoute',
  rootRoute: '_myRootRoute'
})
```

### Available configuration props

| key | type | default value | description
| ------ | ------ | ------ | ------ |
| debugRoute | string (optional) | _debug | change the route used to activate profiler on current request |
| rootRoute | string (optional) | _profiler | change the profiler scopes view and main view route |
| enableCache | boolean (optional) | false | if `true`, then browser cache is enabled. This will not send the `cache-control: no-cache` header. So some requests might not reach the server as they are cached by the browser. Consequently, no new token will be generated and no data is logged for the cached request as the browser is showing data from the response of an old request.
| scopes | ProfilerScope[] (optional) | [] | array of enabled scopes instances. Check [more scopes][URL_more-scopes] |
| streamMiddlewares | RequestHandler[] (optional) | []| array of enabled streamMiddlwares. Check [more scopes][URL_more-scopes]|

## More Scopes

You can import more scopes from the following:

- [axios-scope][URL_scopes_axios]
- [mongoose-scope][URL_scopes_mongoose]

## Demo

[Demo][URL_demo]

[Code base][URL_example_app]

## Known Issues

### Version Mismatch

```sh
Error: The scope {"name":"mongoose","hasToolbarSlot":true,"hasPageView":true} is not instance of ProfilerScope
```

When facing an issue like the above. This means that the failing `scope package` is not using the same `@ghadautopia/express-profiler` which is installed in your project. Please make sure to install the `scope package` which aligns to the same version of `@ghadautopia/express-profiler` used in you project. You can debug the failing `scope package` via going into the project's `node_modules`. Then search for the `scope package` directory and in its `package.json`, you can see the version of the required `@ghadautopia/express-profiler` under the `dependencies` field.

## Advanced

- [Create your own scope][URL_create-scope]

## License

MIT

[//]: # (reference links)

   [URL_symfony]: <https://symfony.com/doc/current/profiler.html>
   [URL_express]: <https://expressjs.com/>
   [URL_scopes_axios]: <https://github.com/ghadautopia/express-profiler-axios>
   [URL_scopes_mongoose]: <https://github.com/ghadautopia/express-profiler-mongoose>
   [URL_create-scope]: <doc/create-scope.md>
   [URL_more-scopes]: <#more-scopes>
   [URL_example_app]: <https://github.com/ghadautopia/express-profiler-example>
   [URL_demo]: <http://express-profiler-example.herokuapp.com/_debug/>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [v1.0.0] - 2021-12-24
### Added
- Intial project

## [v1.0.1] - 2021-12-24
### Changed
- Update readme

## [v1.0.2] - 2021-12-24
### Changed
- Update readme

### Removed
- nodemon from dev dependencies


src/assets/toolbar.ts
import { ToolbarSlotDTO } from "../types/internals";

type AddElement = (data: ToolbarSlotDTO & { url?: string }) => void;
type Toolbar = HTMLElement & { addElement: AddElement };

const tokenLink = document.currentScript?.dataset.tokenLink;
const dataRoute = document.currentScript?.dataset.dataRoute || "";
const svgCollapsed =
  '<svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" width="24" height="24" viewBox="0 0 24 24"><path d="M23 20.168l-8.185-8.187 8.185-8.174-2.832-2.807-8.182 8.179-8.176-8.179-2.81 2.81 8.186 8.196-8.186 8.184 2.81 2.81 8.203-8.192 8.18 8.192z"/></svg>';
const svgExpanded = `<svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" version="1.0" width="28" height="28" viewBox="0 0 628.000000 978.000000" preserveAspectRatio="xMidYMid meet">
<g transform="translate(0.000000,978.000000) scale(0.100000,-0.100000)" stroke="none">
<path d="M2445 9358 c-70 -79 -232 -246 -359 -371 -194 -192 -228 -231 -215 -240 229 -174 413 -348 636 -604 63 -73 119 -133 124 -133 11 0 538 519 641 632 72 79 80 91 70 112 -26 58 -179 244 -298 362 -130 129 -435 384 -460 384 -7 0 -70 -64 -139 -142z"/>
<path d="M2425 7864 c-357 -80 -766 -398 -1006 -784 -111 -180 -193 -389 -224 -579 -23 -136 -16 -349 14 -465 29 -108 96 -245 171 -346 76 -104 250 -276 353 -350 42 -30 76 -58 77 -61 0 -3 -28 -37 -61 -75 -260 -292 -593 -837 -747 -1224 -263 -661 -330 -1345 -186 -1880 168 -622 660 -1192 1264 -1466 314 -141 626 -206 1000 -207 216 0 349 13 485 48 175 44 269 85 660 284 227 116 403 171 605 190 63 6 151 14 195 18 l79 8 70 207 c39 114 68 209 65 212 -2 2 -85 16 -184 30 -178 25 -285 36 -770 76 -646 54 -1063 114 -1341 195 -559 163 -942 457 -1144 879 -109 230 -149 432 -150 756 0 285 44 471 176 734 81 160 165 276 308 425 421 434 1041 751 1960 1001 98 27 182 52 186 57 17 18 270 913 270 956 0 9 -6 17 -14 17 -37 0 -1247 -405 -1641 -549 -76 -28 -150 -51 -163 -51 -29 0 -178 58 -283 110 -210 106 -404 279 -500 447 l-24 42 40 35 c99 89 244 190 338 236 374 181 884 239 1295 145 40 -9 79 -15 86 -12 15 6 286 400 286 416 0 12 -202 166 -295 226 -193 122 -440 219 -710 277 -176 37 -425 48 -540 22z"/>
</g>
</svg>`;

window.onload = () => {
  fetch(dataRoute)
    .then((resp) => {
      return resp.json();
    })
    .then((data) => {
      const toolbar = createToolbar();

      if (!data.length) {
        toolbar.addElement({
          text: "N/A",
          name: "",
        });
      }

      data.forEach((scope: ToolbarSlotDTO) => {
        toolbar.addElement({
          ...scope,
          url: `${tokenLink}/${scope.name}`,
        });
      });

      document.body.appendChild(toolbar);
    });
};

function createToolbar(): Toolbar {
  const iframe = document.createElement("iframe");
  iframe.style.position = "fixed";
  iframe.style.padding = "0";
  iframe.style.border = "none";
  iframe.style.margin = "0";
  iframe.style.bottom = "0";
  iframe.style.right = "0";
  iframe.style.backgroundColor = "#000";
  iframe.style.color = "#FFF";
  iframe.style.width = "100vw";
  iframe.style.fontSize = "14px";
  iframe.style.whiteSpace = "nowrap";
  iframe.style.height = "43px";

  iframe.addEventListener("load", function () {
    const content = iframe.contentDocument || iframe.contentWindow?.document;

    if (content) {
      content.documentElement.style.margin = "0";
      content.documentElement.style.padding = "0";
      const fontLink = document.createElement("link");
      fontLink.href =
        "https://fonts.googleapis.com/css2?family=Outfit:wght@100;200;300;400&display=swap";
      fontLink.rel = "stylesheet";
      content.head.appendChild(fontLink);
      content.body.style.margin = "0";
      content.body.style.padding = "0";
      content.body.style.display = "flex";
      content.body.style.justifyContent = "space-between";
      content.body.style.fontFamily = "'Outfit', sans-serif";
      content.body.appendChild(wrapper);

      const closeEl = document.createElement("section");
      closeEl.style.color = "#FFF";
      closeEl.style.display = "flex";
      closeEl.style.alignItems = "center";
      closeEl.style.justifyContent = "center";
      closeEl.style.height = "100%";
      closeEl.style.width = "43px";
      closeEl.style.cursor = "pointer";
      closeEl.style.flex = "0 0 auto";
      closeEl.setAttribute("title", "toggle toolbar");
      closeEl.innerHTML = svgCollapsed;
      content.body.appendChild(closeEl);

      closeEl.onclick = () => {
        if (wrapper.dataset.collapsed) {
          closeEl.innerHTML = svgCollapsed;
          wrapper.removeAttribute("data-collapsed");
          wrapper.style.overflow = "auto";
          wrapper.style.maxWidth = "";
          iframe.style.maxWidth = "";
        } else {
          closeEl.innerHTML = svgExpanded;
          wrapper.setAttribute("data-collapsed", "1");
          wrapper.style.overflow = "hidden";
          wrapper.style.maxWidth = "0";
          iframe.style.maxWidth = "43px";
        }
      };
    }
  });

  const wrapper = document.createElement("section");
  wrapper.style.display = "flex";
  wrapper.style.height = "100%";
  wrapper.style.width = "100%";
  wrapper.style.overflow = "auto";

  const addElement: AddElement = function ({
    svg,
    text,
    url,
    description,
    color,
  }) {
    const ch = document.createElement("a");
    if (url) ch.setAttribute("href", url);
    ch.setAttribute("target", "_blank");
    ch.setAttribute("title", description || text);
    ch.style.display = "flex";
    ch.style.alignItems = "center";
    ch.style.justifyContent = "center";
    ch.style.height = "100%";
    ch.style.padding = "0 13px";
    ch.style.maxWidth = "80px";
    ch.style.flex = "0 0 auto";
    ch.style.whiteSpace = "no-wrap";
    ch.style.textOverflow = "ellipsis";
    ch.style.overflow = "hidden";
    ch.style.color = "#FFF";
    ch.style.textDecoration = "none";
    ch.style.backgroundColor = color || "#000";

    if (svg) {
      const svgWrapper = document.createElement("div");
      svgWrapper.innerHTML = svg;
      const svgEl = svgWrapper.firstChild as HTMLElement;
      svgEl.style.fill = "currentColor";
      svgEl.style.marginRight = "2px";
      ch.appendChild(svgEl);
    }

    const textEl = document.createTextNode(text);
    ch.appendChild(textEl);

    wrapper.appendChild(ch);
  };

  return Object.assign(iframe, { addElement });
}
